tests/engine/test_validation.py

```diff
@@ -0,0 +1,511 @@
+from rasa.shared.nlu.training_data.training_data import TrainingData
+
+
+class TestComponentWithoutRun(GraphComponent):
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> GraphComponent:
+        return cls()
+
+
+class TestComponentWithRun(TestComponentWithoutRun):
+    def run(self) -> TrainingData:
+        pass
+
+
+def create_test_schema(
+    uses: Type,  # The unspecified type is on purpose to enable testing of invalid cases
+    constructor_name: Text = "create",
+    run_fn: Text = "run",
+    needs: Optional[Dict[Text, Text]] = None,
+    eager: bool = True,
+    parent: Optional[Type[GraphComponent]] = None,
+) -> GraphSchema:
+    parent_node = {}
+    if parent:
+        parent_node = {
+            "parent": SchemaNode(
+                needs={}, uses=parent, constructor_name="create", fn="run", config={}
+            )
+        }
+    # noinspection PyTypeChecker
+    return GraphSchema(
+        {
+            "my_node": SchemaNode(
+                needs=needs or {},
+                uses=uses,
+                eager=eager,
+                constructor_name=constructor_name,
+                fn=run_fn,
+                config={},
+            ),
+            **parent_node,
+        }
+    )
+
+
+def test_graph_component_is_no_graph_component():
+    class MyComponent:
+        def other(self) -> TrainingData:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_component_fn_does_not_exist():
+    schema = create_test_schema(uses=TestComponentWithRun, run_fn="some_fn")
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_output_is_not_fingerprintable_int():
+    class MyComponent(TestComponentWithoutRun):
+        def run(self) -> int:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_predict_graph_output_is_not_fingerprintable():
+    class MyComponent(TestComponentWithoutRun):
+        def run(self) -> int:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    validation.validate(schema, language=None, is_train_graph=False)
+
+
+def test_graph_output_is_not_fingerprintable_any():
+    class MyComponent(TestComponentWithoutRun):
+        def run(self) -> Any:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_output_is_not_fingerprintable_None():
```

Comment:
can we parameterize the `test_graph_output_is_not_fingerprintable_...` ones?

Reply:
unfortunately not. The type annotations are unfortunately not dynamically called. I tried to parametrize types and put that in the type annotations of `MyComponent` but that doesn't work. We could try to dynamically set them via something like `MyComponent.run.__annnotations__["return"] = xy` but that in my opinion makes the tests very unreadable and very hacky.